core/src/main/java/io/ddf/DDFCache.java

```diff
@@ -68,19 +106,18 @@ public synchronized void setDDFName(DDF ddf, String name) throws DDFException {
   }
 
   public synchronized void setDDFUUID(DDF ddf, UUID uuid) throws DDFException {
-    if(this.hasDDF(uuid)) {
+    if(this.containsDDF(uuid)) {
       throw new DDFException(String.format("DDF with uuid %s already exists", uuid));
     } else {
-      //remove old key
       UUID prevUUID = ddf.getUUID();
-      if(prevUUID != null) {
-        mDDFs.remove(prevUUID);
-      }
       ddf.setUUID(uuid);
-      mDDFs.put(uuid, ddf);
-      if(ddf.getUri() != null) {
-        mUris.remove(ddf.getUri());
-        mUris.put(ddf.getUri(), ddf.getUUID());
+      if(prevUUID != null) {
+        mDDFCache.invalidate(prevUUID);
+        mDDFCache.put(uuid, ddf);
+        if(ddf.getUri() != null) {
+          mUris.remove(ddf.getUri());
```

Comment:
@Huandao0812 UUID is right to remove prevUUID then put new.
but Uri is not as Binh said.
```
_mUris.remove(ddf.getUri());_
mUris.put(ddf.getUri(), ddf.getUUID());
```